lib/AST/ASTScopeLookup.cpp

```diff
@@ -152,24 +175,24 @@ ASTScopeImpl::findChildContaining(SourceLoc loc,
 
     bool operator()(const ASTScopeImpl *scope, SourceLoc loc) {
       ASTScopeAssert(scope->checkSourceRangeOfThisASTNode(), "Bad range.");
-      return -1 == ASTScopeImpl::compare(scope->getSourceRangeOfScope(), loc,
-                                         sourceMgr,
+      auto rangeOfScope = scope->getSourceRangeOfScope();
+      loc = translateLocForReplacedRange(sourceMgr, rangeOfScope, loc);
+      return -1 == ASTScopeImpl::compare(rangeOfScope, loc, sourceMgr,
                                          /*ensureDisjoint=*/false);
     }
     bool operator()(SourceLoc loc, const ASTScopeImpl *scope) {
-      ASTScopeAssert(scope->checkSourceRangeOfThisASTNode(), "Bad range.");
-      // Alternatively, we could check that loc < start-of-scope
-      return 0 >= ASTScopeImpl::compare(loc, scope->getSourceRangeOfScope(),
-                                        sourceMgr,
-                                        /*ensureDisjoint=*/false);
+      return !(*this)(scope, loc);
```

Comment:
Could you please double-check this? (Or maybe just reassure me) What about the == case?

Reply:
Thanks for notifying! I think your concern is legitimate (the comparison function should not return `true` for `==` case).
But, it turned out that this function is not called. According to https://en.cppreference.com/w/cpp/algorithm/lower_bound , comparison happens only by `comp(element, value)` , but not `comp(value, element)`. 

> The signature of the predicate function should be equivalent to the following:
> ```
> bool pred(const Type1 &a, const Type2 &b);
> ```
> The type `Type1` must be such that an object of type `ForwardIt` can be dereferenced and then implicitly converted to `Type1`. The type `Type2` must be such that an object of type `T` can be implicitly converted to `Type2`

So I wrote this block to:
```cpp
  auto *const *child = llvm::lower_bound(
      getChildren(), loc,
      [&sourceMgr](const ASTScopeImpl *scope, SourceLoc loc) {
         ...
      });
```

Reply:
Thank you for addressing it. I worry that a future maintainer, (or even myself) will get confused. If the function on line 183 is not called, do we still need to supply it? If so, can we just put an llvm unreachable in it? Otherwise, how about a comment around line 184 explaining away the potential confusion?

Reply:
Hmm, since I removed `operator()(SourceLoc loc, const ASTScopeImpl *scope)` and rewrote `CompareLocs` with a lambda expression, I believe there's no unreachable code anymore.